### PR TITLE
feat: add support for custom context path

### PR DESCRIPTION
**Pull request by tvcsantos**
This PR adds support for custom context path.

Redirection was broken when using a custom context path. It follows the standard using "X-Forwarded-Prefix" header to get the correct context path for redirection.

Thanks for the project.

If you need me to update something feel free to comment ;)

### CODE REVIEW

File: .devcontainer/Dockerfile [hunk 1]
```diff
@@ -31,6 +31,7 @@ WORKDIR /flipt
 
 COPY go.mod go.mod
 COPY go.sum go.sum
+COPY ./core ./core
```
Comment (markphelps) [inline .devcontainer/Dockerfile:34]: thank you! 😄

File: ui/src/components/header/UserProfile.tsx [hunk 2]
```diff
@@ -52,14 +52,28 @@ export default function UserProfile(props: UserProfileProps) {
           metadata[authMethodIssuerKey as keyof typeof metadata];
         logoutURL = `//${logoutURI}`;
       }
+
+      const redirectAddressKey = 'io.flipt.auth.redirect_address';
+      if (metadata[redirectAddressKey as keyof typeof metadata]) {
+        let redirectAddress = metadata[redirectAddressKey as keyof typeof metadata] as string;
+        redirectAddress = redirectAddress?.trim();
```
Comment (markphelps) [inline ui/src/components/header/UserProfile.tsx:59]: if you run `mage ui:fmt` it will fix any ui formatting issues for you automatically

File: internal/server/authn/method/github/server.go [hunk 3]
```diff
@@ -240,9 +259,9 @@ func api(ctx context.Context, token *oauth2.Token, endpoint endpoint, v any) err
 	return json.NewDecoder(resp.Body).Decode(v)
 }
 
-func getUserOrgs(ctx context.Context, token *oauth2.Token) (map[string]bool, error) {
+func getUserOrgs(apiURL string, ctx context.Context, token *oauth2.Token) (map[string]bool, error) {
```
Comment (markphelps) [inline internal/server/authn/method/github/server.go:262]:
```suggestion
func getUserOrgs(ctx context.Context, apiURL string, token *oauth2.Token) (map[string]bool, error) {
```

nit: its customary for `context.Context` to be the first argument

File: internal/server/authn/method/github/server.go [hunk 4]
```diff
@@ -254,9 +273,9 @@ func getUserOrgs(ctx context.Context, token *oauth2.Token) (map[string]bool, err
 	return orgs, nil
 }
 
-func getUserTeamsByOrg(ctx context.Context, token *oauth2.Token) (map[string]map[string]bool, error) {
+func getUserTeamsByOrg(apiURL string, ctx context.Context, token *oauth2.Token) (map[string]map[string]bool, error) {
```
Comment (markphelps) [inline internal/server/authn/method/github/server.go:276]:
```suggestion
func getUserTeamsByOrg(ctx context.Context, apiURL string, token *oauth2.Token) (map[string]map[string]bool, error) {
```
same as above

**Review by markphelps — COMMENTED** [review]
thank you for the PR! will be a great feature to have and say that we now work with GHES!

Just a few minor suggestions.

I want to also loop in @GeorgeMac re the the redirect_address/forward prefix bits as he implemented the majority of the current auth flow

**Comment by markphelps** [comment]
I added the `needs_docs` label just so we don't forget to update the [docs](https://github.com/flipt-io/docs) before creating a new release with this feature included

don't feel like you need to add the docs however, we can do that, unless of course you want to ;)

**Review by GeorgeMac — CHANGES_REQUESTED** [review]
I think this is awesome. Thank you for making this happen!

There is one tiny tidy up and one slight adjustment to building the x forwarded prefix I think might be worth doing.

On the redirect stuff, I had to refresh my memory a bit. I think the change here seems reasonable to me. However, I do think there is another way, involving leaning into that `X-Forwarded-Prefix` some more when setting the redirect location on the ForwardResponseOption.
Im trying to figure out what makes most for us going forward.

The alternative approach would be something like this:

We can get the original HTTP headers off the `ctx`, as GRPC gateway smuggles them in there. However, by default, it only copies across [permenant headers](https://github.com/grpc-ecosystem/grpc-gateway/blob/main/runtime/context.go#L334) as defined in the IANA standard. So we would need to do a little bit of threading ourselves. It is possible to override the header matcher / extend it by wrapping this https://pkg.go.dev/github.com/grpc-ecosystem/grpc-gateway/runtime#DefaultHeaderMatcher to allow through other headers (`X-Forwarded-Prefix` in this case). Then we could get it in the `ForwardResponseOption` using the [`metadata.FromIncoming`](https://pkg.go.dev/google.golang.org/grpc/metadata#FromIncomingContext) function on the context and reading it off the metadata.
Long story short, is we could just redirect to the relative forwarded prefix in this situation and I think it would achieve the same ends? Correct me if I am wrong.

The only upside here is we don't need to encode the redirect address in the authentication metadata each time and have to thread that around. I don't know if anyone has any preferences on that here? (thoughts @markphelps @erka ?)

File: internal/server/authn/method/http.go [hunk 5]
```diff
@@ -94,6 +104,12 @@ func (m Middleware) ForwardResponseOption(ctx context.Context, w http.ResponseWr
 func (m Middleware) Handler(next http.Handler) http.Handler {
 	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
 		prefix, method := path.Split(r.URL.Path)
+		forwardedPrefixValue := r.Header.Get("X-Forwarded-Prefix")
+		forwardedPrefixes := strings.Split(forwardedPrefixValue, ",")
+		for i, prefix := range forwardedPrefixes {
+			forwardedPrefixes[i] = strings.TrimSpace(prefix)
+		}
+		forwardedPrefix := strings.Join(forwardedPrefixes, "")
```
[the remark this answers is not in the document]
Comment (tvcsantos) [reply internal/server/authn/method/http.go:112]: Yes actually I was inspired by Spring :). Ty for the improvement 👍

**Review by GeorgeMac — CHANGES_REQUESTED** [review]
Hey @tvcsantos !

Just got off an offline sync with @markphelps where we touched base on this topic.
We're a little concerned about the context prefix more broadly.
There has been a previously opened issue and some work to attempt to address this:
- https://github.com/flipt-io/flipt/issues/2723
- https://github.com/flipt-io/flipt/pull/2734

The end result we have at the moment, I don't think solves the full story.
It probably works for situations where the UI is served using Vite, which is something we do in development only, not production. So you probably can get most of it working by changing the `base` for Vite, but this is not going to work in the production distributions of Flipt.

I might be mistaken here, maybe you could try building your application as an image and trying out your fix end to end with that instead of the dev environment?

```
mage dagger:run "build:flipt"
```
> This should produce an image you can run locally. The dagger output can be a bit obtuse, but you should see an image URL printed in there somewhere at the end.

But if this doesn't work because none of the paths are prefixed appropriately, I wonder if we should take what you have here and separate it into two pieces of work? One for context path and one for GHE. Then we can get your GHE changes in ASAP and think through the broader context paths issue as its own change. WDYT?

**Comment by tvcsantos** [comment]
> I added the `needs_docs` label just so we don't forget to update the [docs](https://github.com/flipt-io/docs) before creating a new release with this feature included
> 
> don't feel like you need to add the docs however, we can do that, unless of course you want to ;)

I will do the docs if needed no worries there

**Comment by tvcsantos** [comment]
@GeorgeMac I will try to give you feedback on it still this week. But yeah I agree with you maybe we should split in 2 PRs.

I will try to have time tomorrow to split them and then provide feedback regarding the custom path topic.

Thanks for the comments and the collaboration

**Comment by GeorgeMac** [comment]
Thank you for your contributions @tvcsantos 🙌 We definitely want to get these changes in either way :D

**Comment by erka** [comment]
Thank you for your hard work here @tvcsantos. 

@GeorgeMac I've made some updates regarding the custom context path in #3135, specifically involving the X-Forwarded-Prefix header. These changes are intended to help advance this PR, but they might be somewhat bold. I would appreciate your feedback on this.

**Comment by tvcsantos** [comment]
@markphelps @GeorgeMac created a separate PR to deal with the addition of GHE support #3136.

**Comment by tvcsantos** [comment]
Since this thread on the PR is mostly about the context path feature. I will cleanup this PR having only the context path related code, change the title and maintain it open.

**Comment by tvcsantos** [comment]
Regarding the context path fix, actually in order to test before opening the PR I have created a docker image on my GitHub

https://github.com/tvcsantos/flipt/tree/feature/add-self-hosted-v2

```
docker pull ghcr.io/tvcsantos/flipt:feature-add-self-hosted-v2
```

**Comment by tvcsantos** [comment]
I will close this PR since this is being addressed on other #3135
